### ISSUE TITLE
[Enhancement] NLL_loss efficiency

### ISSUE THREAD

**Issue by NatureGeorge** [yd-yin/RotationLaplace#1]
Hi! Your work is interesting. I previously adopt David's [work](https://github.com/Davmo049/Public_prob_orientation_estimation_with_matrix_fisher_distributions) which implemented the Matrix Fisher distribution (MF) for regressing rotations. And recently I tried switching to your proposed Rotation Laplace distribution (RL). 

But I found that computing the negative log-likelihood of RL is not as efficient as MF, in both time and memory. I suspect it results from the size of the grid that you set to compute. The shape of the grid is `(36864, 3, 3)`. So it inevitably slows down the computation of `NLL_loss`.

It would make RL more applicable if the computation is optimized.

**Comment by yd-yin** [comment]
Hi, thanks for your interest, and thanks for your suggestion!

Yes, David's work tried some mathematical methods to analytically approximate the normalization factor, which is pretty nice. However, the analytical approximation is not universal to all distributions, and the computation of the factor is always a hard problem in terms of building a new distribution.

In our experiments, we found discretization with 36864 grids is generally a very accurate approximation and is not a bottleneck in training (see Table 6 in supp) with a 3090 GPU. What GPU do you use? An option can be to try decreasing the number of samples to 4.6k or 0.6k and see if the performance is still good. 

We're happy to see if we can find some nice way to make the computation more efficient.

**Comment by NatureGeorge** [comment]
Appreciate your kind reply.

I tested RL on 3070 (8G, laptop) and V100 (32G). In both cases, the memory consumption ran high immediately and ran out of CUDA memory within several steps, which I did not encounter when using MF. Each of the tests adapts the same setting as I use before for MF. I typically would deal with thousands of rotations in one mini-batch cause I am not doing pose estimation but another heavy task.

I would try switching to the lower grid sizes.
